src/radiation/moments.cpp

```diff
@@ -15,14 +15,75 @@
 #include <kokkos_abstraction.hpp>
 #include <utils/error_checking.hpp>
 
+#include <singularity-eos/eos/eos.hpp>
+
+#include "phoebus_utils/programming_utils.hpp"
+#include "phoebus_utils/root_find.hpp"
 #include "radiation/closure.hpp"
 #include "radiation/closure_m1.hpp"
+#include "radiation/closure_mocmc.hpp"
 #include "radiation/local_three_geometry.hpp"
 #include "radiation/radiation.hpp"
 #include "reconstruction.hpp"
 
 namespace radiation {
 
+using namespace singularity::neutrinos;
+using singularity::EOS;
+
+class InteractionTResidual {
+ public:
+  KOKKOS_FUNCTION
+  InteractionTResidual(const EOS &eos, const Opacity &opacity,
+                       const MeanOpacity &mean_opacity, const Real &rho, const Real &ug0,
+                       const Real &Ye, const Real J0[3], const int &nspec,
+                       const RadiationType species[3], const Real &scattering_fraction,
+                       const Real &dtau)
+      : eos_(eos), opacity_(opacity), mean_opacity_(mean_opacity), rho_(rho), ug0_(ug0),
+        Ye_(Ye), nspec_(nspec), scattering_fraction_(scattering_fraction), dtau_(dtau) {
+    for (int ispec = 0; ispec < nspec; ++ispec) {
+      J0_[ispec] = J0[ispec];
+      species_[ispec] = species[ispec];
+    }
+  }
+
+  KOKKOS_INLINE_FUNCTION
+  Real operator()(const Real T) {
+    Real lambda[2] = {Ye_, 0.0};
+
+    Real J0_tot = 0.;
+    Real dJ_tot = 0.;
+
+    for (int ispec = 0; ispec < nspec_; ++ispec) {
+      J0_tot += J0_[ispec];
+      const Real kappa =
+          (1. - scattering_fraction_) *
+          mean_opacity_.RosselandMeanAbsorptionCoefficient(rho_, T, Ye_, species_[ispec]);
+
+      const Real JBB = opacity_.EnergyDensityFromTemperature(T, species_[ispec]);
+
+      dJ_tot += (J0_[ispec] + dtau_ * kappa * JBB) / (1. + dtau_ * kappa) - J0_[ispec];
+    }
+
+    const Real ug1 = rho_ * eos_.InternalEnergyFromDensityTemperature(rho_, T, lambda);
+
+    return ((ug1 - ug0_) + (dJ_tot)) / (ug0_ + J0_tot);
```

Comment:
@lroberts36 Here is the implicit-in-fluid-temperature residual. I solve `ug(T^n+1) - ug(T^n) = ur^n+1 - ur^n` in the fluid frame. Then after I get `T^n+1`, I use `LinearSourceUpdate` to get `dE` and `dF` and update the conserved radiation and fluid variables as usual.

Reply:
I like this, it is a good solution.